### ISSUE TITLE
Kill signal is not propagated for Singularity container

### ISSUE THREAD

**Issue by stevekm** [nextflow-io/nextflow#1561]
## Bug report 

In the `.command.run` file, the default `on_term` handler for TERM, INT, USR1, and  USR2 signals looks like this:

```
on_term() {
    set +e
    [[ "$pid" ]] && nxf_kill $pid
}
```

However when we run Nextflow with Singularity enabled, our `on_term` keeps coming out like this:

```
on_term() {
    set +e
}
```

It looks like `on_term` gets populated here;

https://github.com/nextflow-io/nextflow/blob/8138b3c4b7c1c22a4d3b5f47f85a64077f34b6c5/modules/nextflow/src/main/resources/nextflow/executor/command-run.txt#L79-L82
```
on_term() {
    set +e
    {{kill_cmd}}
}
```

based on the values returned here:
https://github.com/nextflow-io/nextflow/blob/21088ad3e5890d4ecdb0f8e17b75f6d5b9e99ba6/modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy#L216-L221
```
            binding.kill_cmd = containerBuilder.getKillCommand()
        }
        else {
            binding.container_boxid = null
            binding.container_helpers = null
            binding.kill_cmd = '[[ "$pid" ]] && nxf_kill $pid'
        }

```

The default value for `containerBuilder.getKillCommand()` appears to be `null`;
https://github.com/nextflow-io/nextflow/blob/b2cfb66d637d046477c30d8900a1e037db3479f2/modules/nextflow/src/main/groovy/nextflow/container/ContainerBuilder.groovy#L122
```
    String getKillCommand() { return null }
```

However it appears that for Docker and PodMan, a custom kill command is set;
https://github.com/nextflow-io/nextflow/search?q=containerBuilder.getKillCommand&unscoped_q=containerBuilder.getKillCommand

However there does not appear to be anything for Singularity. 

This is breaking our pipeline (as per https://github.com/mskcc/tempo/issues/736) because LSF is giving off USR2 signals when our jobs go over the allocated resources, but these signals are resulting in some programs to exit with a 0 exit code, and this value gets saved in `.exitcode`. Since the expected `output` file was produced (in this case, broken truncated .bam files), Nextflow does not realize that the task failed and propagates the broken files downstream in the pipeline, causing silent errors and other problems.

Is there a reason why Singularity does not have a `kill_cmd` set? It seems like the lack of this could be contributing to our difficulties handling USR2 signals on LSF

### Environment 

* Nextflow version: 19.10.0
* Java version: 8
* Operating system: CentOS 7
* Bash version: 4.2.46

**Comment by stevekm** [comment]
@gongyixiao

**Comment by pditommaso** [comment]
Thanks for the detailed analysis, this is very useful. I think the `getKillCommand` is null because Singularity doesn't have a specific stop/kill command. 

However, being an in-process execution it should be enough like any other process and therefore it should fallback to `'[[ "$pid" ]] && nxf_kill $pid'`

**Comment by stevekm** [comment]
Thanks. Do you think it would be safe for us to include the 'default' `nxf_kill` inside our `process.beforeScript` block in order to restore this kind of functionality as a hold-over until a change gets made in Nextflow? Maybe something like this?

```
process.beforeScript = """
on_term() {
    set +e
    [[ "$pid" ]] && nxf_kill $pid
}
"""
```

**Comment by gongyixiao** [comment]
> Do you think it would be safe for us to include the 'default' nxf_kill invocation inside our process.beforeScript block in order to restore this kind of functionality as a hold-over until a change gets made in Nextflow?

I have gave this solution a try. It doesn't seem to override the `on_term () { set +e }` in `.command.run`

> Thanks for the detailed analysis, this is very useful. I think the getKillCommand is null because Singularity doesn't have a specific stop/kill command.

How about `singularity stop instance` or `singularity oci kill`? Sorry I'm not a singularity pro. It's just a quick google search result.

**Comment by pditommaso** [comment]
The problem is that singularity 2.x has no `stop` command. IMO the usual process kill should work. You may want to give it a try.

**Comment by stevekm** [comment]
> IMO the usual process kill should work. You may want to give it a try.

We had issues with this, it did not work consistently. Something about issues with the way that the Singularity process is pushed to background on our LSF system; @gongyixiao had more details on it. In the end, the fix that seemed to help the most looked like this:

```
  beforeScript = "catch_USR2 () { echo 'caught USR2 signal'; set +e; false; on_exit ; } ; trap catch_USR2 USR2"
```

Not sure if this was a "just us" type of issue on our HPC or something that could be affecting other systems as well. I suspect that on other systems, the regular 'nxf_kill' invocation would work, though I dont really have a way to test it.

**Comment by pditommaso** [comment]
This was included in version `20.07.1`. Feel free to re-open if the problem still arise.

**Comment by lescai** [comment]
Hi guys, running version 20.11.0-edge and experiencing similar issue.
Running on SLURM version 18.08.4-626_cm8.2.x86_64
And using Singularity 3.3.0

```
eb-17 19:49:20.844 [Task monitor] ERROR nextflow.processor.TaskProcessor - Error executing process > 'RNASEQ:DESEQ2_QC_STAR_SALMON'
  Process `RNASEQ:DESEQ2_QC_STAR_SALMON` terminated for an unknown reason -- Likely it has been terminated by the external system
Feb-17 19:49:25.779 [Task monitor] DEBUG nextflow.executor.GridTaskHandler - Failed to get exit status for process TaskHandler[jobId: 157490; id: 2233; name: RNASEQ:DESEQ2_QC_SALMON; status: RUNNING; exit: -; error: -; workDir: [-removed-] started: 1613558620721; exited: -; ] -- exitStatusReadTimeoutMillis: 270000; delta: 274997
```

for a job that has been killed due to time limit
`slurmstepd: error: *** JOB 157490 ON node19 CANCELLED AT 2021-02-17T19:43:41 DUE TO TIME LIMIT ***`

checking with 
```
sjobexitmod -l 157490
       JobID    Account   NNodes        NodeList      State ExitCode DerivedExitCode        Comment 
------------ ---------- -------- --------------- ---------- -------- --------------- -------------- 
157490                         1          node19    TIMEOUT      0:0             0:0                
```

and like in earlier comments with 
```
sacct -j 157490 --format=User,JobID,Jobname,partition,state,ExitCode
     User        JobID    JobName  Partition      State ExitCode 
--------- ------------ ---------- ---------- ---------- -------- 
   lescai 157490       nf-RNASEQ+    compute    TIMEOUT      0:0 
          157490.batch      batch             CANCELLED      0:9 
```

Exit code reported in NFtower is -
